### PR TITLE
[BUGFIX beta] Don't use prototype extensions (arrayContent{Did,Will}Change, removeAt)

### PR DESCRIPTION
**Pull request by btecu**
Do not rely on prototype extensions for `removeAt`.
Related to #9269 and #10899.

### CODE REVIEW

File: packages/ember-runtime/lib/mixins/mutable_array.js [hunk 1]
```diff
@@ -128,9 +145,9 @@ export default Mixin.create(EmberArray, MutableEnumerable, {
     ```javascript
     var colors = ['red', 'green', 'blue', 'yellow', 'orange'];
 
-    colors.removeAt(0);     // ['green', 'blue', 'yellow', 'orange']
-    colors.removeAt(2, 2);  // ['green', 'blue']
-    colors.removeAt(4, 2);  // Error: Index out of range
+    removeAt(colors, 0);     // ['green', 'blue', 'yellow', 'orange']
```
Comment (rwjblue) [inline packages/ember-runtime/lib/mixins/mutable_array.js]: These examples are showing how to use the mutable array mixin itself, I would think that we want to leave them as `colors.removeAt(2, 2)`.

**Comment by rwjblue** [comment]
I'm not 100% sure on this.  It seems like the right direction, but I wouldn't expect so many changes in tests that were already using things extended by `NativeArray` (i.e. `Ember.A()`).  We still support `Ember.A().removeAt(...)` and friends, so we should continue to test them.

**Comment by krisselden** [comment]
This is much like moving away from global API internally.  In general, the tests migrated to the modules API but we should have tests that test we export the API onto `Ember`.

**Comment by stefanpenner** [comment]
nice!

**Comment by krisselden** [comment]
we should probably get this in ASAP as it is likely to bit quickly with the churn in glimmer 2 PRs

**Comment by krisselden** [comment]
@rwjblue I think NativeArray could have a test but that any other tests not testing NativeArray directly should be using the standalone methods.

**Comment by btecu** [comment]
I've added one test and update those comments.

**Comment by btecu** [comment]
Rebased

File: packages/ember-runtime/lib/system/array_proxy.js [hunk 2]
```diff
@@ -33,6 +35,40 @@ var EMPTY = [];
 
 function K() { return this; }
 
+export function removeAt(array, start, len) {
+  if ('number' === typeof start) {
```
Comment (krisselden) [inline packages/ember-runtime/lib/system/array_proxy.js]: why extract this? it doesn't let you fire the events for this operation without using Ember.A().

Comment (krisselden) [inline packages/ember-runtime/lib/system/array_proxy.js]: the first argument is still required to be an ember array.

This isn't the equivalent of Ember.set(obj, key, val) where the obj can be a pojo.

Comment (btecu) [inline packages/ember-runtime/lib/system/array_proxy.js]: @krisselden to be able to call `removeAt(array, 1, 2);` in some other files such as `arranged_content_test.js`, although there might not be any advantage over `array.removeAt(1, 2);`.

Comment (krisselden) [inline packages/ember-runtime/lib/system/array_proxy.js]: it if requires it to be an array, there isn't a point to the standalone, there is only value migrating to extend prototypes being false by default if Ember isn't doing Ember.A() all over the place, which is very slow.

Comment (btecu) [inline packages/ember-runtime/lib/system/array_proxy.js]: That makes sense. I've reverted that chage.

**Comment by btecu** [comment]
Rebased.

**Comment by rwjblue** [comment]
Thanks @btecu, sorry for letting this linger so long.

**Comment by workmanw** [comment]
I'm not sure where or how to direct this feedback, so I'm going to do so here. If there is a better place for this discussion please let me know.

These commits are problematic for [ember-data-model-fragments](https://github.com/lytics/ember-data-model-fragments). The root of the problem is that because `replace` (and others) no longer call `this.arrayContentDidChange(...)`, instead calling `(arrayContentDidChange(this, ...)`, we are unable to override this function in our subclasses array. See here: [array/stateful.js#L178-L190](https://github.com/lytics/ember-data-model-fragments/blob/master/addon/array/stateful.js#L178-L190).

Ultimately I acknowledge that this an model-fragments issue. We were overriding a public API and now the way it's invoked internally from ember has change. So I don't think this is exactly API breakage, but I can definitely see this as problematic for anyone else in a similar situation.

**Comment by stefanpenner** [comment]
@workmanw that sounds like a breaking change to me

**Comment by workmanw** [comment]
@stefanpenner If that's the case, I'd like to get this fixed up before ember 2.8 ships so it doesn't become an upgrade blocker for us, or anyone else. I can submit a PR to revert part of this commit so that internally ember will call `this.arrayContentDidChange(...)` instead of `arrayContentDidChange(this, ...)` if you think that's the right direction.

**Comment by btecu** [comment]
Seems unlikely that breaking the overriding of an api can be considered a breaking change.

**Comment by stefanpenner** [comment]
@btecu it is a public API, although maybe this is a gray area?

@workmanw can you open an issue so we can track and decide if it is an issue or not?

**Comment by Serabe** [comment]
@stefanpenner it seems so. Though a public API, it is a method to call to notify about the change, not to be notified about it. Interesting issue!

**Comment by workmanw** [comment]
@stefanpenner Yea, absolutely.

**Pull request merged**